src/deploy/index.js

```diff
@@ -1,38 +1,85 @@
+const _ = require('lodash');
+const util = require('util');
 const fs = require('fs');
+const path = require('path');
 const Listr = require('listr');
-const Observable = require('zen-observable');
-const {spawn} = require('child_process');
-
-const main = async () => {
-  const cfg = {};
-  const lockFile = JSON.parse(fs.readFileSync('services.lock'));
-
-  const publisher = new Listr(
-    Object.keys(lockFile).map(serviceName => {
-      const service = Object.assign({}, {name: serviceName}, lockFile[serviceName]);
-      return {
-        title: service.name,
-        task: () => new Listr([
-          {
-            title: 'Docker push',
-            task: async () => {
-              const cmd = spawn('docker', ['push', service.tag]);
-              return new Observable(observer => {
-                observer.next('waiting...');
-                cmd.on('error', observer.error);
-                cmd.on('close', observer.complete);
-                cmd.stdout.on('data', observer.next);
-                cmd.stderr.on('data', observer.next);
-              });
-            },
-          },
-        ]),
-      };
-    }),
-    {concurrent: 1}
-  );
-
-  await publisher.run();
+const {ClusterSpec} = require('../formats/cluster-spec');
+const config = require('typed-env-config');
+const jsone = require('json-e');
+const mkdirp = util.promisify(require('mkdirp'));
+const rimraf = util.promisify(require('rimraf'));
+
+class Deploy {
+  constructor(input, output, infraInfo) {
+    this.input = input;
+    this.output = output;
+    this.infraInfo = infraInfo;
+
+    this.spec = null;
+    this.cfg = null;
+  }
+
+  _readInputs() {
+    this.spec = new ClusterSpec(this.input);
+    this.infra = JSON.parse(fs.readFileSync(this.infraInfo));
+    this.cfg = config({
+      files: [
+        'deploy-config.yml',
+        'user-deploy-config.yml',
+      ],
+      env:      process.env,
+    });
+  }
+
+  _renderResources() {
+    const contextFunctions = {
+      builtService: name => {
+        const service = _.find(this.spec.build.services, {name});
+        if (!service) {
+          throw new Error(`builtService: unkonwn service ${name}`);
+        }
+        return service;
+      },
+    };
+
+    const context = {
+      cfg: this.cfg,
+      infra: this.infra,
+      build: this.spec.build,
+      ...contextFunctions,
+    };
+    this.resources = jsone(this.spec.deploy, context).resources;
```

Comment:
Oh, this is a neat hack. 

I guess maybe it's not even really a hack, but still cool.